### PR TITLE
Add certificate rotation to teleport openssh join oneshot command 

### PR DESCRIPTION
**Pull request by lxea**
Will have a follow up pr that updates the scripts and tries to execute certificate rotation

### CODE REVIEW

File: tool/teleport/common/agentless.go [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	return true, nil
+}
+
+// openSSHInitialJoin
+func (a *agentless) openSSHInitialJoin(ctx context.Context, clf config.CommandLineFlags) error {
```
Comment (russjones) [inline tool/teleport/common/agentless.go:467]: I'd suggest some refactoring around this function because I think it mixes two things: initial joining and rotation for a joined node). You can still re-use a bunch of common logic, but you just have to break it up a bit different.

What do you think about something like below? It makes it very easy for the reader to understand what happens during each phase of rotation.

```go
func (a *agentless) init(ctx context.Context, clf config.CommandLineFlags) error {
  // Connect to auth, register this node, and get a certificate bundle back.
  certs, err := a.register()

  // Reconnect to auth with the new identity.
  client, err := a.connectIdentity(certs)

  // Update key and certificates that will be used by OpenSSH on disk.
  err = a.updateKeysAndConfig(certs)

  // Restart OpenSSH.
  err = a.restart()

  // Register this node with Teleport so it can show up in Teleport inventory.
  err = registerServer()

  return nil
}


func (a *agentless) rotate(ctx context.Context, clf config.CommandLineFlags) error {
  // Read existing identity from disk.
  oldCerts, err := readIdentity()

  // Use existing identity to connect to auth.
  client, err := connectIdentity(oldCerts)

  // Request new host certificates.
  certs, err := client.GenerateHostCerts(oldCerts)

  // Update key and certificates that will be used by OpenSSH on disk.
  err = a.updateKeysAndConfig(certs)

  // Restart OpenSSH.
  err = a.restart()

  // Register this node with Teleport so it can show up in Teleport inventory.
  err = registerServer()

  return nil
}
```

I would suggest a similar approach to the rollback function.

File: tool/teleport/common/agentless.go [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+			HostID:               a.uuid,
+			NodeName:             a.hostname,
+			Role:                 types.RoleNode,
+			AdditionalPrincipals: a.principals,
+			PublicTLSKey:         tlsPublicKey,
+			PublicSSHKey:         sshPublicKey,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
```
Comment (russjones) [inline tool/teleport/common/agentless.go:507]: In this block, you're never re-connecting to auth with the new identity. Could that cause problems?

File: tool/teleport/common/agentless.go [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+
+	openSSHCA, err := a.getOpenSSHCA(ctx, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	hostSSHCA, hostTLSCA, err := a.getHostCA(ctx, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
Comment (russjones) [inline tool/teleport/common/agentless.go:534]: Why are you fetching these again? Don't you get the CAs as part of `certs` when you register or generate host certificates?

File: tool/teleport/common/agentless.go (hunk 1 above)
Comment (ibeckermayer) [reply tool/teleport/common/agentless.go:467]: In `init`, what's the difference between the time you "register this node" on the first line and the time you do it on the last line?

File: tool/teleport/common/agentless.go [hunk 4]
```diff
@@ -0,0 +1,725 @@
...
+	"golang.org/x/crypto/ssh"
+)
+
+const sshdBinary = "sshd"
+
+const sshdConfigSectionModificationHeader = "# Created by 'teleport join openssh', do not edit"
+
+var (
+	// agentlessSSHDConfigPath is the path to write teleport specific SSHD config options
+	agentlessSSHDConfigPath = "/etc/teleport/agentless_sshd.conf"
```
Comment (r0mant) [inline tool/teleport/common/agentless.go:54]: I would remove the word agentless from everywhere. It's an implementation detail.
```suggestion
	agentlessSSHDConfigPath = "/etc/teleport/sshd.conf"
```

File: tool/teleport/common/agentless.go [hunk 5]
```diff
@@ -0,0 +1,725 @@
...
+	// agentlessSSHDConfig is Include directive added to the sshd_config file
+	agentlessSSHDConfigInclude = "Include " + agentlessSSHDConfigPath
+	// agentlessKeys is the path to write agentless openssh keys
+	agentlessKeysDir = "/etc/teleport/agentless"
```
Comment (r0mant) [inline tool/teleport/common/agentless.go:58]:
```suggestion
	agentlessKeysDir = "/etc/teleport/openssh"
```

File: tool/teleport/common/agentless.go [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+	// agentlessKeys is the path to write agentless openssh keys for
+	// use on rollback
+	agentlessKeysBackupDir = "/etc/teleport/agentless_backup"
```
Comment (r0mant) [inline tool/teleport/common/agentless.go:61]:
```suggestion
	agentlessKeysBackupDir = "/etc/teleport/openssh_backup"
```

File: tool/teleport/common/agentless.go [hunk 7]
```diff
@@ -0,0 +1,725 @@
...
+
+func readKeys(clf config.CommandLineFlags) (privateKey, publicKey, tlsPublicKey []byte, certs *proto.Certs, err error) {
+	privateKey, err = os.ReadFile(filepath.Join(clf.OpenSSHKeysPath, teleportKey))
+	if err != nil {
+		return nil, nil, nil, nil, trace.ConvertSystemError(err)
+	}
+
+	publicKey, err = os.ReadFile(filepath.Join(clf.OpenSSHKeysPath, teleportCert))
+	if err != nil {
+		return nil, nil, nil, nil, trace.Wrap(err)
```
Comment (r0mant) [inline tool/teleport/common/agentless.go:133]: Use `trace.ConvertSystemError(err)` here and everywhere below in this method.

File: tool/teleport/common/teleport.go [hunk 8]
```diff
@@ -422,10 +409,12 @@ func Run(options Options) (app *kingpin.Application, executedCommand string, con
...
 	joinOpenSSH.Flag("openssh-config", "Path to the OpenSSH config file").Default("/etc/ssh/sshd_config").StringVar(&ccf.OpenSSHConfigPath)
 	joinOpenSSH.Flag("openssh-keys-path", "Path to the place teleport keys and certs").Default(agentlessKeysDir).StringVar(&ccf.OpenSSHKeysPath)
+	joinOpenSSH.Flag("openssh-keys-backup-path", "Path to the place teleport keys and certs should be backed up to after certificate rotation").Default(agentlessKeysBackupDir).StringVar(&ccf.OpenSSHKeysBackupPath)
 	joinOpenSSH.Flag("restart-sshd", "Restart OpenSSH").Default("true").BoolVar(&ccf.RestartOpenSSH)
 	joinOpenSSH.Flag("insecure", "Insecure mode disables certificate validation").BoolVar(&ccf.InsecureMode)
 	joinOpenSSH.Flag("additional-principals", "Comma separated list of host names the node can be accessed by").StringVar(&ccf.AdditionalPrincipals)
 	joinOpenSSH.Flag("debug", "Enable verbose logging to stderr.").Short('d').BoolVar(&ccf.Debug)
+	joinOpenSSH.Flag("rotate", "Optionally Rotate the CA certs (update_servers, rollback)").EnumVar(&ccf.RotateOpenSSH, types.RotationPhaseUpdateServers, types.RotationPhaseRollback)
```
Comment (r0mant) [inline tool/teleport/common/teleport.go:417]: This flag help message is confusing. I would rephrase to `Optionally specify the CA rotation phase to execute. Can be one of: update_servers (xxx) or rollback (xxx)` and fill in "xxx" with brief description of what will happen during this phase.

File: tool/teleport/common/agentless.go
[the remark this answers is not in the document]
Comment (r0mant) [reply tool/teleport/common/agentless.go:1]: I agree, this file also feels like a weird mix of CA rotation control logic and filesystem and other utility functions.

I would avoid the word "agentless" but I would create a package `lib/openssh` and put most of this logic there exposing some interface you can call from `tool/teleport`, like your `Agentless`. Although I'm not really a fan of this name tbh. Maybe just call it `Rotation` or something?

For the structure, I'd do:

- `lib/openssh/rotate.go` --> control logic for the rotation, basically your "agentless" struct.
- `lib/opebssh/sshd.go` --> utility methods for interacting with OpenSSH, e.g. restarting it, updating its config, etc. I'd also expose it as an interface.
- `lib/openssh/keys.go` --> methods for generating/writing keys you have.

**Review by r0mant — COMMENTED** [review]
First pass, just comments about code organization for now, will do another pass to review the rotation logic in more detail.

File: tool/teleport/common/agentless.go
[the remark this answers is not in the document]
Comment (espadolini) [reply tool/teleport/common/agentless.go:1]: Why not just reuse the same process storage that normal teleport uses, rather than figure out some new way to store credentials and host CAs (to communicate with the teleport cluster)? That way the problem goes from implementing a whole new mechanism to just spinning up half of a `TeleportProcess` to call that one function that refreshes host credentials, and then just updating whatever files are needed for openssh to work (in addition to the "natural" host credential updates).

File: lib/config/configuration.go [hunk 9]
```diff
@@ -181,10 +181,14 @@ type CommandLineFlags struct {
 	OpenSSHConfigPath string
 	// OpenSSHKeysPath is the path to write teleport keys and certs into
 	OpenSSHKeysPath string
+	// OpenSSHKeysBackupPath is the path to write teleport keys and certs into after rotation
+	OpenSSHKeysBackupPath string
 	// AdditionalPrincipals are a list of extra principals to include when generating host keys.
 	AdditionalPrincipals string
 	// RestartOpenSSH indicates whether openssh should be restarted or not.
 	RestartOpenSSH bool
+	// RotateOpenSSH indicates that the OpenSSHCA certs should be rotated
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/config/configuration.go:190]: Why is this up to the command line arguments and not up to the state of the cluster? The host ssh certificate (that's the SSH half of the host Node credentials) to use depends on the state of the cluster's `host` `cert_authority`, and since we're going to be merely consumers of the `openssh` `cert_authority`, all we need to do there is fetch it - because we have to do that anyway to know if it has changed - and then update the list of trusted user cert authorities for opensshd with all the ssh keys we find in there. I don't see how the user should be allowed to specify which "kind" of rotation we should do.

File: tool/teleport/common/agentless.go
[the remark this answers is not in the document]
Comment (r0mant) [reply tool/teleport/common/agentless.go:1]: I was thinking along the same lines. @lxea Is there a reason we can't reuse existing code Teleport itself uses to fetch credentials with the cluster to simplify things?

For example, we could initialize the `TeleportProcess` as mentioned, and then use [LocalRegister](https://github.com/gravitational/teleport/blob/master/lib/auth/register.go#L53) to mimic what Teleport process would do during [firstTimeConnect](https://github.com/gravitational/teleport/blob/master/lib/service/connect.go#L553) and then pass the fetched credentials to some OpenSSH "service" to place them where needed and restart sshd.

@espadolini Is this similar to what you had in mind?

[the remark this answers is not in the document]
Comment (espadolini) [reply tool/teleport/common/agentless.go:1]: Yeah, or even just call the actual methods in `TeleportProcess` (or extract whatever code we need in standalone methods, which can only improve the organization of the `TeleportProcess` methods); we need the equivalent of `connect`, `syncRotationState`/`syncServiceRotationState`, and then some custom code that calls `GetCertAuthority` for the openssh CA.

File: lib/openssh/sshd.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+	if err := sshdConfigTmp.Sync(); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	if err := b.sshd.checkConfig(sshdConfigTmp.Name()); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.Rename(sshdConfigTmp.Name(), u.TeleportSSHDConfigPath); err != nil {
```
Comment (espadolini) [inline lib/openssh/sshd.go:98]: This only works if the `sshdConfigTmp` happened to be created on the same filesystem as `u.TeleportSSHDConfigPath`, which is not guaranteed (often times `/tmp/` will be a tmpfs, for instance). You should create the temporary file in the same directory as `TeleportSSHDConfigPath` and defer its deletion blindly (so that if we exit out before the file is renamed, the temporary file is deleted).

File: lib/openssh/sshd.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+	contents, err := os.ReadFile(sshdConfigPath)
+	if err != nil {
+		return false, trace.ConvertSystemError(err)
+	}
+	return !strings.Contains(string(contents), fileContains), nil
+}
+
+func fmtSSHDConfigUpdate(openSSHCAPath string, hostKeyPaths, hostCertPaths []string) string {
+	return fmt.Sprintf(`%s
+TrustedUserCaKeys %s
```
Comment (espadolini) [inline lib/openssh/sshd.go:201]: nit
```suggestion
TrustedUserCAKeys %s
```

File: lib/openssh/sshd.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+HostKey %s
+HostCertificate %s
+`,
+		sshdConfigSectionModificationHeader,
+		openSSHCAPath,
+		strings.Join(hostKeyPaths, " "),
+		strings.Join(hostCertPaths, " "),
```
Comment (espadolini) [inline lib/openssh/sshd.go:208]: I don't think `HostKey` takes multiple paths like that; even if it did, `sshd` will only ever present a single key per signature algo, I'm afraid, so multiple `ssh-cert-rsa-v01@openssh.com` certs are going to be useless. 😢

File: lib/service/connect.go [hunk 13]
```diff
@@ -684,6 +688,115 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+	}
+	conn, err := process.WaitForConnector(SSHIdentityEvent, nil)
+	if conn == nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	ctx := context.Background()
+
+	status, err := process.syncRotationStateAndBroadcast(conn)
```
Comment (espadolini) [inline lib/service/connect.go:708]: Shouldn't this be `process.syncRotationState`?

File: lib/service/connect.go [hunk 14]
```diff
@@ -684,6 +688,115 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	fmt.Println(status)
+
+	id, err := process.storage.ReadIdentity(auth.IdentityCurrent, types.RoleInstance)
```
Comment (espadolini) [inline lib/service/connect.go:713]: Why the `Instance` credentials and not the `Node` credentials?

File: lib/openssh/sshd.go [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+		}
+	}
+
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportOpenSSHCA), caKeyBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	return nil
+}
+
+func TryCreateDefaultAgentlesKeysDir(agentlessKeysPath string) error {
```
Comment (espadolini) [inline lib/openssh/sshd.go:130]:
```suggestion
func TryCreateDefaultAgentlessKeysDir(agentlessKeysPath string) error {
```

File: lib/openssh/sshd.go [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+func prefixSSHDConfig(sshdConfigPath, config string) error {
+	contents, err := os.ReadFile(sshdConfigPath)
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	line := append([]byte(config), byte('\n'))
+	contents = append(line, contents...)
+
+	err = os.WriteFile(sshdConfigPath, contents, 0644)
```
Comment (espadolini) [inline lib/openssh/sshd.go:183]: If this fails halfway through we hose the host's `sshd_config`, which isn't ideal; we should do an atomic file update here as well.

File: lib/openssh/sshd.go [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+
+func TryCreateDefaultAgentlesKeysDir(agentlessKeysPath string) error {
+	_, err := os.Stat(agentlessKeysPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return trace.Wrap(os.MkdirAll(agentlessKeysPath, 0700))
+		}
+		return trace.Wrap(err)
+	}
+	return nil
```
Comment (espadolini) [inline lib/openssh/sshd.go:138]: Isn't this just `os.MkdirAll(agentlessKeysPath, 0o700)` but more complicated?

(also the explicit octal `0o700` is preferrable to the implicit one)

File: lib/service/connect.go [hunk 18]
```diff
@@ -684,6 +688,115 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+		return trace.Wrap(err)
+	}
+
+	cas, err := conn.Client.GetCertAuthorities(ctx, types.OpenSSHCA, false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	keysDir := process.Config.OpenSSH.KeysDir
+	if err := openssh.TryCreateDefaultAgentlesKeysDir(keysDir); err != nil {
```
Comment (espadolini) [inline lib/service/connect.go:724]: I think that `openssh.WriteKeys` could just try to create the directory instead of requiring the caller to do this extra step.

File: lib/service/connect.go [hunk 19]
```diff
@@ -684,6 +687,114 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+	if err := process.storage.WriteState(types.RoleNode, *state); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := registerServer(process.Config, ctx, conn.Client, mostRecentRotation); err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		process.Shutdown(ctx)
```
Comment (espadolini) [inline lib/service/connect.go:767]: Took me a sec to figure out that this is all within `RegisterCriticalFunc` so if we exit with a non-nil error the process dies anyway; please add a comment about that.

File: lib/service/connect.go [hunk 20]
```diff
@@ -684,6 +687,114 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+		if err := process.Close(); err != nil {
```
Comment (espadolini) [inline lib/service/connect.go:768]: I don't think we need a `process.Close` after `process.Shutdown`, do we?

**Comment by r0mant** [comment]
@lxea Is this ready for another review?

File: lib/service/connect.go [hunk 21]
```diff
@@ -684,6 +687,108 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return trace.Wrap(err)
+	}
+	conn, err := process.WaitForConnector(SSHIdentityEvent, nil)
+	if conn == nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close()
+
+	status, err := process.syncRotationState(conn)
```
Comment (ibeckermayer) [inline lib/service/connect.go:706]: I notice that this is otherwise called from `syncRotationStateAndBroadcast`, can it ever be a problem that we sync but don't "broadcast" here?

File: lib/service/connect.go [hunk 22]
```diff
@@ -684,6 +687,108 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
 	return connector, nil
 }
 
+func (process *TeleportProcess) initOpenSSH() {
+	process.RegisterWithAuthServer(types.RoleNode, SSHIdentityEvent)
+	process.SSHD = openssh.NewSSHD(process.Config.OpenSSH.RestartCommand)
+	process.RegisterCriticalFunc("common.rotate_openssh2", process.syncOpenSSHRotationState)
```
Comment (ibeckermayer) [inline lib/service/connect.go:693]: Reasoning behind the "common" prefix here? I would expect something more like `"openssh2.rotate"` based on other names.

File: lib/service/connect.go [hunk 23]
```diff
@@ -684,6 +687,114 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+
+	if err := registerServer(process.Config, ctx, conn.Client, mostRecentRotation); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// if any of the above exits with non nil error, the process is
+	// shut down as it is run via RegisterCriticalFunction, so we
+	// manually shut down here as we dont want teleport to remain
+	// running after
+	go func() {
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply lib/service/connect.go:761]: Can you add an inline comment noting that? I was also confused.

File: lib/service/connect.go (hunk 20 above)
Comment (ibeckermayer) [reply lib/service/connect.go:763]: Do we understand the difference between these two functions? Why is one used some places and the other in others? Can/should they be consolidated?

File: lib/openssh/sshd.go [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+	}
+
+	if restart {
+		if err := b.sshd.restart(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/openssh/sshd.go:132]: IDK about this, it's fairly typical to use single-letter identifiers for the receiver of a function, isn't it?

File: lib/service/connect.go (hunk 21 above)
Comment (espadolini) [reply lib/service/connect.go:706]: The only relevant thing that `syncRotationStateAndBroadcast` does is broadcast a `TeleportReloadEvent` if the host credentials have changed and Teleport should restart, but we are doing this one-shot thing, so we shouldn't actually reload.

File: lib/openssh/sshd.go [hunk 25]
```diff
@@ -0,0 +1,243 @@
...
+	// SSHDConfigPath is the path to write teleport specific SSHD config options
+	SSHDConfigPath = filepath.Join(defaults.DataDir, "sshd.conf")
+	// SSHDKeysDir is the path to the Teleport openssh keys
+	SSHDKeysDir = filepath.Join(defaults.DataDir, "openssh")
+
+	// sshdConfigInclude is Include directive added to the sshd_config file
+	sshdConfigInclude = "Include " + SSHDConfigPath
+)
+
+const DefualtRestartCommand = "systemctl restart sshd"
```
Comment (espadolini) [inline lib/openssh/sshd.go:45]:
```suggestion
const DefaultRestartCommand = "systemctl restart sshd"
```

File: lib/openssh/sshd.go [hunk 26]
```diff
@@ -0,0 +1,243 @@
...
+
+const (
+	// TeleportKey is the name the OpenSSH private key
+	TeleportKey = "ssh_host_teleport_key"
+	// TeleportKey is the name the OpenSSH cert
+	TeleportCert = "ssh-host-teleport-cert.pub"
```
Comment (espadolini) [inline lib/openssh/sshd.go:51]: I'm not sure if there's some convention regarding where to store host certificates, but if we extrapolate from `.ssh/id_rsa`:

```suggestion
	TeleportCert = "ssh_host_teleport_key-cert.pub"
```

or maybe just

```suggestion
	TeleportCert = TeleportKey + "-cert.pub"
```

File: lib/openssh/sshd.go [hunk 27]
```diff
@@ -0,0 +1,246 @@
...
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportCert), id.CertBytes, 0600); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	var caKeyBytes []byte
+	for _, ca := range cas {
+		for _, key := range ca.GetTrustedSSHKeyPairs() {
+			caKeyBytes = append(caKeyBytes, key.PublicKey...)
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/openssh/sshd.go:150]: Maybe we should strip trailing newlines and then add one, then? I think that the openssh authorized keys parser will simply skip over empty lines, but I'm not sure.

File: lib/openssh/sshd.go [hunk 28]
```diff
@@ -0,0 +1,243 @@
...
+	return nil
+}
+
+// WriteKeys writes the OpenSSH keys and CA from the Identity and the
+// OpenSSH CA to disk for the OpenSSH daemon to use
+func WriteKeys(keysdir string, id *auth.Identity, cas []types.CertAuthority) error {
+	if err := os.MkdirAll(keysdir, 0o700); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportKey), id.KeyBytes, 0600); err != nil {
```
Comment (espadolini) [inline lib/openssh/sshd.go:140]: I feel like all of these should use `writeTempAndRename` just so we don't make a machine unconnectable if the power is cut at exactly the wrong time.

File: lib/service/connect.go (hunk 20 above)
Comment (espadolini) [reply lib/service/connect.go:767]: I wonder if it's some awkward thing caused by using the `ExitContext` as the shutdown context here; what happens if you pass `context.Background()` to `process.Shutdown` instead?

File: lib/service/connect.go [hunk 29]
```diff
@@ -684,6 +687,112 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+		return trace.Wrap(err)
+	}
+
+	state, err := process.storage.GetState(types.RoleNode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	mostRecentRotation := state.Spec.Rotation
+	if !status.needsReload {
```
Comment (espadolini) [inline lib/service/connect.go:745]: Even if we got `needsReload` because the host CA was rotated we might still have an even more recent openssh CA, right?

**Review by espadolini — COMMENTED** [review]
I'm not a very big fan of how "openssh mode" is kinda shoehorned in as a very special case, but we can just clean it up if and when we end up adding a different one-shot mode.

File: lib/openssh/sshd.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply lib/openssh/sshd.go]: True, I was mostly perturbed by `b` being used.

File: lib/openssh/sshd.go [hunk 30]
```diff
@@ -0,0 +1,254 @@
...
+	}
+	if err := sshdConfigTmp.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := checkFunc(tmpName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := os.Rename(sshdConfigTmp.Name(), configPath); err != nil {
```
Comment (r0mant) [inline lib/openssh/sshd.go:227]:
```suggestion
	if err := os.Rename(tmpName, configPath); err != nil {
```

File: lib/openssh/sshd.go [hunk 31]
```diff
@@ -0,0 +1,254 @@
...
+	cmd := exec.Command(sshdBinary, "-t")
+	if err := cmd.Run(); err != nil {
+		output, outErr := cmd.CombinedOutput()
+		if outErr != nil {
+			return trace.Wrap(trace.NewAggregate(err, outErr), "invalid sshd config file, failed to get `sshd -t` output")
+		}
+		return trace.Wrap(err, "invalid sshd config file %q, not writing", string(output))
+	}
+
+	cmd = exec.Command("/bin/sh", "-c", b.restartCmd)
```
Comment (r0mant) [inline lib/openssh/sshd.go:199]: Why do we need to call this via shell?

File: lib/config/configuration.go [hunk 32]
```diff
@@ -1963,8 +1969,10 @@ func Configure(clf *CommandLineFlags, cfg *servicecfg.Config, legacyAppFlags boo
 		cfg.Databases.Databases = append(cfg.Databases.Databases, db)
 	}
 
-	if err = ApplyFileConfig(fileConf, cfg); err != nil {
-		return trace.Wrap(err)
+	if !cfg.OpenSSH.Enabled {
```
Comment (r0mant) [inline lib/config/configuration.go:1972]: One thing I'm not really a fan of is that we're updating "Configure" method to have a special behavior if OpenSSH "mode" is enabled, I think Edoardo raised a similar concern.

Instead of changing this function to check `cfg.OpenSSH.Enabled` in several places and updating the config "from within" using "applyOpenSSHConfig" below, can we keep this function as-is, and just have a dedicated "ConfigureOpenSSH" method which would construct config specifically for OpenSSH mode (basicaly, do what "applyOpenSSHConfig" does). Cause you don't need probably 90% of what's in this function for OpenSSH (like, databases, apps and other services).

**Review by r0mant — COMMENTED** [review]
This mostly looks ok to me now. In addition to comments left by other folks, just a couple more about config generation, and plus need to add test coverage. After that it should be good to go.

File: lib/config/configuration.go [hunk 33]
```diff
@@ -2141,6 +2145,68 @@ func Configure(clf *CommandLineFlags, cfg *servicecfg.Config, legacyAppFlags boo
...
+		cfg.SetToken(clf.AuthToken)
+	}
+
+	applyOpenSSHConfig(cfg, clf)
+
+	return nil
+}
+
+func applyOpenSSHConfig(cfg *servicecfg.Config, clf *CommandLineFlags) error {
+	log.Debugf("Disabling all services, only the Teleport openssh service can run during the `teleport openssh join` command")
```
Comment (r0mant) [inline lib/config/configuration.go:2171]:
```suggestion
	log.Debugf("Disabling all services, only the Teleport OpenSSH service can run during the `teleport openssh join` command")
```

File: lib/openssh/sshd.go (hunk 31 above)
Comment (r0mant) [reply lib/openssh/sshd.go:208]: Ah I didn't realize this command is being passed from the config and is a single string. I guess that's fine as-is.

File: lib/service/connect.go [hunk 34]
```diff
@@ -684,6 +688,97 @@ func (process *TeleportProcess) firstTimeConnect(role types.SystemRole) (*Connec
...
+		if err := process.Close(); err != nil {
+			process.log.Errorf("Error closing process: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+func registerServer(a *servicecfg.Config, ctx context.Context, client auth.ClientI) error {
+
```
Comment (r0mant) [inline lib/service/connect.go:759]:
```suggestion
```

**Review by r0mant — APPROVED** [review]
lgtm

@lxea Can you please also write an integration test for this? Right now it's very lightly tested so I'm concerned there will be regressions.

We can do it in a follow-up PR. The test would configure/start this new `openssh_service` and make sure it fetches proper credentials from the cluster, places them where needed and calls the provided restart command.

File: lib/service/servicecfg/openssh.go [hunk 35]
```diff
@@ -0,0 +1,37 @@
...
+	RestartSSHD bool
+	// RestartCommand is the command to use when restarting sshd
+	RestartCommand string
+	// AdditionalPrincipals is a list of additional principals to be included.
+	AdditionalPrincipals []string
+	// InstanceAddr is the connectable address of the OpenSSh instance.
+	InstanceAddr string
+	// ProxyServer is the address of the teleport proxy.
+	ProxyServer *utils.NetAddr
+	// Labels are lebels to set on the instance.
```
Comment (espadolini) [inline lib/service/servicecfg/openssh.go:35]:
```suggestion
	// Labels are labels to set on the instance.
```

File: lib/service/servicecfg/openssh.go [hunk 36]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+type OpenSSHConfig struct {
+	Enabled bool
+	// SSHDConfigPath is the path to the OpenSSH config file.
+	SSHDConfigPath string
+	// RestartSSHD is true if sshd should be restarted after config updates.
+	RestartSSHD bool
+	// RestartCommand is the command to use when restarting sshd
```
Comment (espadolini) [inline lib/service/servicecfg/openssh.go:27]:
```suggestion
	// RestartCommand is the command to use when restarting sshd.
```

File: lib/service/connect.go (hunk 20 above)
Comment (espadolini) [reply lib/service/connect.go]: Isn't the `GracefulExitContext` cancelled at the very beginning of `Shutdown`?

Comment (espadolini) [reply lib/service/connect.go]: Yeah, this is equivalent of passing in a context to `Shutdown` and immediately closing it, as `(*LocalSupervisor).BroadcastEvent` will close the graceful exit context as soon as it receives the `TeleportExitEvent`.

What do the logs show if you just pass `context.Background()` to `process.Shutdown()`, which is what you'd effectively get with an external signal? The supervisor should periodically show a list of services that are waiting to exit.

**Review by espadolini — COMMENTED** [review]
LGTM other than figuring out what's up with the shutdown.

File: lib/service/connect.go (hunk 20 above)
Comment (espadolini) [reply lib/service/connect.go]: The easiest solution to what I think is actually an oversight in the interaction between `WaitAndReload` and shutdown might be to do
```go
ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
// (or just context.WithCancel, but having a timeout is actually more correct here)
defer cancel()
process.Shutdown(ctx)
```
as canceling the context passed to `Shutdown` should also result in closing the `process.exitContext`, which should unblock `WaitForSignal`.

**Review by espadolini — APPROVED** [review]
